video/out/wayland_common.c

```diff
@@ -2722,6 +2770,35 @@ void vo_wayland_handle_scale(struct vo_wayland_state *wl)
                                 lround(mp_rect_h(wl->geometry) / wl->scaling_factor));
 }
 
+bool vo_wayland_valid_format(struct vo_wayland_state *wl, uint32_t drm_format, uint64_t modifier)
+{
+#if HAVE_DRM
+    // Tranches are grouped by preference and the first tranche is at the end of
+    // the list. It doesn't really matter for us since we search everything
+    // anyways, but might as well start from the most preferred tranche.
+    struct vo_wayland_tranche *tranche;
+    wl_list_for_each_reverse(tranche, &wl->tranche_list, link) {
+        bool supported_compositor_format = false;
+        struct drm_format *formats = tranche->compositor_formats;
+        for (int i = 0; i < tranche->num_compositor_formats; ++i) {
+            if (formats[i].format != drm_format)
+                continue;
+            if (modifier == formats[i].modifier && modifier != DRM_FORMAT_MOD_INVALID)
+                return true;
+            supported_compositor_format = true;
+        }
+
+        if (supported_compositor_format && tranche->planar_formats) {
+            for (int i = 0; i < tranche->num_planar_formats; i++) {
+                if (drm_format == tranche->planar_formats[i])
+                    return true;
+            }
+        }
```

Comment:
Using `planar` here is confusing since a planar format usually means a YCbCr format. But in this case planar formats are those that are supported by a DRM plane.

In any case, this is incorrect. IIRC, this means that, if the compositor supports this format with any modifier, then mpv will assume that it supports it with all modifiers. But this is not true. For example, vulkan and opengl often have distinct sets of modifiers and neither is a subset of the other. In particular, vulkan never supports the implicit modifier. This will get mpv killed.

Reply:
It's only supposed to be true with the invalid modifier but it is indeed incorrect.